Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -142,37 +168,89 @@ public override void OnActionReceived(float[] vectorAction)
         bpDict[forearmR].SetJointStrength(vectorAction[++i]);
     }
 
+    void UpdateOrientationCube()
+    {
+        //FACING DIR
+        m_WalkDir = target.position - m_OrientationCube.transform.position;
+        m_WalkDir.y = 0; //flatten dir on the y
+        m_WalkDirLookRot = Quaternion.LookRotation(m_WalkDir); //get our look rot to the target
+        
+        //UPDATE ORIENTATION CUBE POS & ROT
+        m_OrientationCube.transform.position = hips.position;
+        m_OrientationCube.transform.rotation = m_WalkDirLookRot;
+    }
+    
     void FixedUpdate()
     {
+        if (detectTargets)
+        {
+            foreach (var bodyPart in m_JdController.bodyPartsDict.Values)
+            {
+                if (bodyPart.targetContact && bodyPart.targetContact.touchingTarget)
+                {
+                    TouchedTarget();
+                }
+            }
+        }
+        
+        UpdateOrientationCube();
+        
         // Set reward for this step according to mixture of the following elements.
         // a. Velocity alignment with goal direction.
         // b. Rotation alignment with goal direction.
         // c. Encourage head height.
-        // d. Discourage head movement.
-        m_DirToTarget = target.position - m_JdController.bodyPartsDict[hips].rb.position;
         AddReward(
-            +0.03f * Vector3.Dot(m_DirToTarget.normalized, m_JdController.bodyPartsDict[hips].rb.velocity)
-            + 0.01f * Vector3.Dot(m_DirToTarget.normalized, hips.forward)
-            + 0.02f * (head.position.y - hips.position.y)
-            - 0.01f * Vector3.Distance(m_JdController.bodyPartsDict[head].rb.velocity,
-                m_JdController.bodyPartsDict[hips].rb.velocity)
+            +0.02f * Vector3.Dot(m_OrientationCube.transform.forward,
+                Vector3.ClampMagnitude(m_JdController.bodyPartsDict[hips].rb.velocity, maximumWalkingSpeed))
+            + 0.01f * Vector3.Dot(m_OrientationCube.transform.forward, head.forward)
+            + 0.005f * (head.position.y - footL.position.y)
+            + 0.005f * (head.position.y - footR.position.y)
         );
     }
 
     /// <summary>
-    /// Loop over body parts and reset them to initial conditions.
+    /// Agent touched the target
     /// </summary>
-    public override void OnEpisodeBegin()
+    public void TouchedTarget()
     {
-        if (m_DirToTarget != Vector3.zero)
+        AddReward(1f);
+        if (respawnTargetWhenTouched)
         {
-            transform.rotation = Quaternion.LookRotation(m_DirToTarget);
+            GetRandomTargetPos();
         }
+    }
+
+    /// <summary>
+    /// Moves target to a random position within specified radius.
+    /// </summary>
+    public void GetRandomTargetPos()
```

Comment:
This is not a `Get` Rename to `MoveTargetToRandomPosition()`
I would advise against the Agent moving the target, I think the target should be told it was touched and then  handle its own motion.

Reply:
agreed. i think there should be a universal target controller script. we'd need it to work in multiple situations so I think we need a design discussion around this

Reply:
fixed